packages/flutter/lib/src/cupertino/sheet_route.dart

```diff
@@ -0,0 +1,254 @@
+    final double paddingTop = math.max(10, MediaQuery.of(context)?.padding.top ?? 0);
+    return CupertinoUserInterfaceLevel(
+      data: CupertinoUserInterfaceLevelData.elevated,
+      child: Builder(
+        builder: (BuildContext context) => Padding(
+          padding: EdgeInsets.only(top: _kPreviousRouteVisibleOffset + paddingTop),
+          child: Container(
+            clipBehavior: Clip.hardEdge,
+            decoration: BoxDecoration(
+              borderRadius: BorderRadius.vertical(top: topRadius),
+              color: backgroundColor ??
+                  CupertinoColors.systemBackground.resolveFrom(
+                    context,
+                    nullOk: true,
+                  ),
+            ),
+            child: MediaQuery.removePadding(
+              context: context,
+              removeTop: true,
+              child: child,
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+class CupertinoSheetRoute<T> extends SheetRoute<T> {
+  CupertinoSheetRoute({
+    required WidgetBuilder builder,
+    List<double>? stops,
+    double initialStop = 1,
+    RouteSettings? settings,
+    Color? backgroundColor,
+  }) : super(
+          builder: (BuildContext context) {
+            return _CupertinoSheetDecorationBuilder(
+              child: Builder(
+                builder: (BuildContext context) {
+                  return builder(context);
+                },
+              ),
+              backgroundColor: backgroundColor,
+              topRadius: _kCupertinoSheetTopRadius,
+            );
+          },
+          settings: settings,
+          animationCurve: _kCupertinoSheetCurve,
+          stops: stops,
+          initialStop: initialStop,
+        );
+
+  @override
+  bool get bounceAtTop => true;
+
+  @override
+  bool get draggable => true;
+
+  @override
+  bool get expanded => true;
+
+  @override
+  Color? get barrierColor => _kBarrierColor;
+
+  @override
+  bool get barrierDismissible => false;
+
+  @override
+  Widget buildTransitions(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    final double topPadding = MediaQuery.of(context)!.padding.top;
+    final double topOffset = math.max(_kSheetMinimalOffset, topPadding);
+    return AnimatedBuilder(
+      animation: secondaryAnimation,
+      child: child,
+      builder: (BuildContext context, Widget? child) {
+        final double progress = secondaryAnimation.value;
+        final double scale = 1 - progress / 10;
+        final double distanceWithScale = (topOffset + _kPreviousRouteVisibleOffset) * 0.9;
+        final Offset offset = Offset(0, progress * (topOffset - distanceWithScale));
+        return Transform.translate(
+          offset: offset,
+          child: Transform.scale(
+            scale: scale,
+            child: child,
+            alignment: Alignment.topCenter,
+          ),
```

Comment:
it might be possible to use something like SlideTransition and ScaleTransition rather than an AnimatedBuilder, in principle those will be faster because they can skip the rebuild/relayout and only repaint.

Reply:
(they don't currently, but the point is that if we improve them later, it'd improve this code too)